packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +46,73 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
+    [handleSignTransactions.pending]: (state) => ({
+        ...state,
+        status: SIGN_STATUS.IN_PROGRESS
+    }),
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
+        const retryTxDirection = error.message.includes('Exceeded the prepaid gas')
+            ? RETRY_TX.INCREASE
+            : error.message.includes('TotalPrepaidGasExceeded')
+                ? RETRY_TX.DECREASE
+                : undefined;
+        
+        const tryRetryTx = retryTxDirection && !state.transactions.some((t) => 
+            t.actions && t.actions.some((a) => a.functionCall && a.functionCall.gas && (
+                (
+                    state.retryTxDirection === RETRY_TX.INCREASE 
+                    && a.functionCall.gas.gt(new BN(RETRY_TX.GAS.MAX))
+                ) || (
+                    state.retryTxDirection === RETRY_TX.DECREASE 
+                    && a.functionCall.gas.lte(new BN(RETRY_TX.GAS.MIN))
+                )
+            ))
+        );
```

Comment:
> Aren't these logic checks backwards?

It working well. I prepared it this way (by adding `!` before `some`) so we can base it on the positive value (`tryRetryTx` instead of `stopRetryTx`).

> Also, we should be checking if increasing or decreasing the functionCall.gas amount by DIFF would cause it to exceed the min/max, not check just its current value, otherwise we will overshoot and increase the value past the max threshold; we should clamp the functionCall.gas at <= RETRY_TX.GAS.MAX and => RETRY_TX.GAS.MIN.

Good catch. Done.